### ISSUE TITLE
Core/Kuberay

### ISSUE THREAD

**Issue by defuyun** [ray-project/ray#37434]
### Description

Add a way to timeout remote tasks on pending machine.

we are using GKE spot instance and the machine may not always be available. we want to fallback to regular instance if we are waiting too long to provision the machine.

have some options to timeout due to not able to schedule e.g.

`ray.remote(time_out_pending_machine=600s)`

also we tend to hit the max ray replica so probably don't want to timeout if it's already using max machine. (or have some way for us to tell if it's unable to schedule or hit max replica)

**Comment by rkooo567** [comment]
Could be related issues; https://github.com/ray-project/ray/issues/33954
https://github.com/ray-project/ray/issues/18916

**Comment by defuyun** [comment]
yes similar to https://github.com/ray-project/ray/issues/33954. the proposed solution is a bit tricky since
1. we have tasks running for 2hr+ and harder to decide on a good timeout period, and we don't want to wait 2hr to know we have no more machines when we could have already falled back to a different machine type and finished the task.
2. we generally hit the max replica count so I don't want to keep on timing out and resubmitting the task into the queue

**Comment by larrylian** [comment]
I thank`ray.cancel` API can solve this issue. 
```
ref = func.options(num_cpus=1).remote()
time_start = time.now()
....
....
try :
   ray.get(ref, timeout=5)
except TimeoutError:
  if (time.now - time_start) > 60 * 60 * 2:
       ray.cancel(ref)
```

**Comment by rkooo567** [comment]
^ I think that's the solution I also recommended to this user. @defuyun can you also mention if this is sufficient, and if not, why is that?

**Comment by defuyun** [comment]
thanks for replying. I think I might need some help to understand how this is solving the issue. is the code snippet meant to sit in a while loop. so we timeout every 5 seconds and we are suppose to check if it's a scheduling issue on our side. e.g.

```
ref = func.options(num_cpus=1).remote()
time_start = time.now()
while True:
  try :
     ray.get(ref, timeout=5)
  except TimeoutError:
     # somehow check for if this is a scheduling issue.
     if scheduling_issue:
        ray.cancel(ref)
        func.options(resources={"new_machine": 1}).remote()
```

**Comment by rkooo567** [comment]
I see. so you want to avoid canceling the task if it is "already scheduled"?

**Comment by defuyun** [comment]
yeah, if it's already scheduled and running then I don't want to cancel it.
